### ISSUE TITLE
Console is not completing the path whit host and port when Base uri set relative

### ISSUE THREAD

**Issue by eleonoraortega** [mulesoft/api-console#502]
When Base Uri is relative, APikit is giving the path to the console correctly. 
The problem is that the console is not completing the path with the host and the port, therefore it fails. 
Same problem when there was no "baseUri" defined inside the RAML document.

Vogel: what about cases if there isn't a "baseUri"? Should we ignore that and only report the error that this is not a valid URL?

Pawel: the console can't know what base URL should be in this case. Assuming that the protocol and domain is the same is only partially right because it won't cover all use cases. 
In this case it would be an error. The console can't guess missing data because that will lead to more issues. That's why manual setting is better fit here.

**Comment by jarrodek** [comment]
I still don't know how to approach this.
I'm against setting protocol, host and port from current URI as it may lead to more issues when the relative port (for some reason) have been set like this by purpose.
It could be a configuration option for the console but if that's the case then it should be done in RAML definition because is requires the same user attention.

Currently the console renders error message when the base URL is missing and it is up to the user to provide correct information.

**Comment by jarrodek** [comment]
Also, in case of hosting applications that controls the data passed to the console, it should be up to this application to update the URL to appropriate for current context value. The console is as generic as it can be so it can serve more than one purpose giving different implementations ability to shape the way how console behaves.

**Comment by svacas** [comment]
as relative URIs are allowed as baseURIs in RAML I think it makes sense for the console to use the request host and port by default (the user can always override it). This allows accessing the console from different interfaces and reverse proxies without the need of rewriting the baseUri in the raml files

**Comment by jarrodek** [comment]
@svacas pleas, mind my previous comment about setting this value by hosting application. This would resolve your issue, would not create new issues.
The console don't know anything about any proxy and it is up to hosting application to handle proxy traffic properly on behalf of the console (as described in the [docs](https://github.com/mulesoft/api-console/blob/master/docs/cors.md#handling-http-request-by-the-hosting-website--application)).

**Comment by eleonoraortega** [comment]
Hi @jarrodek @svacas Any update on this?

**Comment by weickmanna** [comment]
This would be a very useful feature.

**Comment by jarrodek** [comment]
I can propose a setting option for the api-console elements `baseUri` which will be used to update the URL if url received from selected method is a relative path.
This would be available to set via HTML tag, node builder and the cli tool.
How does it sounds?

**Comment by weickmanna** [comment]
Sounds good

**Comment by svacas** [comment]
if you can provide an example showing how to set it up once implemented it would be great

**Comment by jarrodek** [comment]
@svacas 
For the HTML element it would be:

```html
<api-console base-uri="http://api.domain.com/base/"></api-console>
```
For the [api-console-builder](https://github.com/mulesoft-labs/api-console-builder) module it would be: 

```javascript
const builder = require('api-console-builder');
builder({
  dest: 'build',
  raml: 'path/to/api.raml',
  baseUri: 'http://api.domain.com/base/'
})
```

And finally CLI tool:

```bash
$ api-console build --base-uri "http://api.domain.com/base/" path/to/api.raml
```

In all cases the console would detect if the master URL starts with any supported protocols and if it doesn't that it would apply the `baseUri` property to the URL editor (producing valid URL).

**Comment by jarrodek** [comment]
**Update of the node example**

Node module has a special option to set attributes on the console element. This should be:

```javascript
const builder = require('api-console-builder');
const attributes = [{
   'base-uri': 'http://api.domain.com/base/'
}];
builder({
  dest: 'build',
  raml: 'path/to/api.raml'',
  attributes: attributes
})
```

**Comment by weickmanna** [comment]
@jarrodek

> "In all cases the console would detect if the master URL starts with any supported protocols and if it doesn't that it would apply the baseUri property to the URL editor (producing valid URL)."

Does it mean that if I don't have this base-uri parameter set, and I have a baseUri in my .raml like "/api/{version}", the console would complete this to http://myHost:myPort/api/myVersion/?

**Comment by jarrodek** [comment]
@weickmanna Not really. Without setting the `baseUri` the console would do nothing and the URL in ypur example would be `/api/{version}` + selected endpoint.

The way how transport behaves that different matter. So the XHR and Fetch will use the same authority part as console's when the URL is relative. But this is true if the request is made by the XHR / Fetch in the app. If the user has API console extension loaded then the request is proxied by it and it looses context of current domain. The same for proxies. 

Summarizing. The console currently and event with this option available would do nothing and default behavior for a web application applies (requests are made to the same authority part as current protocol, domain and port when relative URL is used). 
The behavior would change only if you'd set the `baseUri` and the console would apply base URI to relative URLs (so proxied requests would always point to the same resource).

**Comment by weickmanna** [comment]
@jarrodek any advice on how I could accomplish such behaviour?
We would like to use the console both during development, when running against localhost, and on production, when running against our server URL. Currently, we always have to enter the base URL manually then, because we cannot set it at api.raml already ...

**Comment by jarrodek** [comment]
@weickmanna See my comment above: https://github.com/mulesoft/api-console/issues/502#issuecomment-321076404

**Comment by jarrodek** [comment]
This issue will be resolved with the `4.1.0` release. Once I wrote more detailed tests (tomorrow) I'll do the release.
